src/dbnode/persist/fs/merger.go

```diff
@@ -0,0 +1,255 @@
+		multiIterPool = m.multiIterPool
+		identPool     = m.identPool
+		encoderPool   = m.encoderPool
+
+		nsID       = fileID.Namespace
+		shard      = fileID.Shard
+		startTime  = fileID.BlockStart
+		blockSize  = nsOpts.RetentionOptions().BlockSize()
+		blockStart = xtime.ToUnixNano(startTime)
+		openOpts   = DataReaderOpenOptions{
+			Identifier: FileSetFileIdentifier{
+				Namespace:  nsID,
+				Shard:      shard,
+				BlockStart: startTime,
+			},
+			FileSetType: persist.FileSetFlushType,
+		}
+	)
+
+	if err := reader.Open(openOpts); err != nil {
+		multiErr = multiErr.Add(err)
+		return multiErr.FinalError()
+	}
+	defer reader.Close()
+
+	nsMd, err := namespace.NewMetadata(nsID, nsOpts)
+	if err != nil {
+		return err
+	}
+	prepareOpts := persist.DataPrepareOptions{
+		NamespaceMetadata: nsMd,
+		Shard:             shard,
+		BlockStart:        startTime,
+		DeleteIfExists:    false,
+	}
+	prepared, err := flushPreparer.PrepareData(prepareOpts)
+	if err != nil {
+		multiErr = multiErr.Add(err)
+		return multiErr.FinalError()
+	}
+
+	// Writing data for a block is done in two stages. The first stage is
+	// to loop through series on disk and merge it with what's in the merge
+	// target. Looping through disk in the first stage is done intentionally to
+	// read disk sequentially to optimize for spinning disk access.
+	// The second stage is to persist the rest of the series in the
+	// merge target that were not persisted in the first stage.
+
+	// There will only be one BlockReader slice since we're working within one
+	// block here.
+	brs := make([][]xio.BlockReader, 0, 1)
+	// There will likely be at least two BlockReaders - one for disk data and
+	// one for data from the merge target.
+	br := make([]xio.BlockReader, 0, 2)
+
+	// It's safe to share these between iterations and just reset them each time
+	// because the series gets persisted each loop, so previous iterations'
+	// reader and iterator will never be needed.
+	segReader := srPool.Get()
+	defer segReader.Finalize()
+	multiIter := multiIterPool.Get()
+	defer multiIter.Close()
+
+	// First stage: loop through series on disk.
+	for id, tagsIter, data, _, err := reader.Read(); err != io.EOF; id, tagsIter, data, _, err = reader.Read() {
+		if err != nil {
+			multiErr = multiErr.Add(err)
+			return multiErr.FinalError()
+		}
+
+		// Reset BlockReaders.
+		brs = brs[:0]
+		br = br[:0]
+		br = append(br, blockReaderFromData(data, segReader, startTime, blockSize))
+
+		// Check if this series is in memory (and thus requires merging).
+		encoded, hasData, err := mergeWith.Read(id, blockStart, nsCtx)
+		if err != nil {
+			multiErr = multiErr.Add(err)
+			return multiErr.FinalError()
+		}
+		if hasData {
+			br = append(br, encoded...)
+		}
+		brs = append(brs, br)
+
+		multiIter.ResetSliceOfSlices(xio.NewReaderSliceOfSlicesFromBlockReadersIterator(brs), nsCtx.Schema)
+
+		// tagsIter is never nil.
+		tags, err := convert.TagsFromTagsIter(id, tagsIter, identPool)
```

Comment:
The place where `convert.TagsFromTagsIter` is called is only for series IDs that came from the (disk) reader. The lifecycle for those IDs are contained to within the merger (created from the `Read()` and then finalized after closing the preparer), so using that function should be safe, right?

Agreed on the `iter -> tags -> iter` comment. Created https://github.com/m3db/m3/issues/1675.

Reply:
@justinjc Yes I think with the restructuring you've done this is now safe since the tags will be valid as long as the IDs are valid, and the IDs are valid for the duration of the file writing. Can you add a comment explaining this now?

Reply:
Sure thing.